### PR TITLE
Bypass slow builder for common case for ImmutableSet union

### PR DESCRIPTION
**Pull request by gabbard**

### CODE REVIEW

File: tests/test_immutableset.py [hunk 1]
```diff
@@ -209,7 +209,16 @@ def test_union(self):
...
         s3 = immutableset(["b", "c"])
         self.assertEqual(s1, s2.union(s3))
+        self.assertEqual(s1, s2.union(s3, check_top_type_matches=str))
         self.assertEqual(s1, s2 | s3)
+        # check union with plain iterable
+        self.assertEqual(s1, s2.union(["b", "c"]))
+
+        with self.assertRaises(RuntimeError):
+            # right hand provokes type failure
+            self.assertEqual(s1, s2.union([1, 2, 3], check_top_type_matches=str))
```
Comment (berquist) [inline tests/test_immutableset.py]: For both of these, do you need the assertion in order to trigger the error?
```suggestion
            s2.union([1, 2, 3], check_top_type_matches=str)
```

File: immutablecollections/_immutableset.py [hunk 2]
```diff
@@ -246,12 +247,28 @@ def union(
...
-            .add_all(other)
-            .build()
-        )
+        if check_top_type_matches:
+            for (side_name, side) in (("left", self), ("right", other)):
+                items_not_matching = [
+                    x for x in side if not isinstance(x, check_top_type_matches)
+                ]
+                if items_not_matching:
+                    raise RuntimeError(
```
Comment (berquist) [inline immutablecollections/_immutableset.py]: I think `TypeError` is a better choice, according to https://docs.python.org/3/library/exceptions.html#TypeError.

File: tests/test_immutableset.py (hunk 1 above)
Comment (gabbard) [reply tests/test_immutableset.py]: Fixed in 7283c8cfe075d9d72e0eb061e80ffd62ae2bc22e

File: immutablecollections/_immutableset.py (hunk 2 above)
Comment (gabbard) [reply immutablecollections/_immutableset.py]: Fixed in 463cbbe50706a6ee48aa83abf6aa5c6e3699715e

File: tests/test_immutableset.py [hunk 3]
```diff
@@ -209,7 +209,16 @@ def test_union(self):
         s2 = immutableset(["a", "c"])
         s3 = immutableset(["b", "c"])
         self.assertEqual(s1, s2.union(s3))
+        self.assertEqual(s1, s2.union(s3, check_top_type_matches=str))
         self.assertEqual(s1, s2 | s3)
+        # check union with plain iterable
+        self.assertEqual(s1, s2.union(["b", "c"]))
+
+        with self.assertRaises(RuntimeError):
```
Comment (berquist) [inline tests/test_immutableset.py]: CI is failing because of
```suggestion
        with self.assertRaises(TypeError):
```

**Pull request merged**